### PR TITLE
Fix issue with Array in the Inline templates

### PR DESCRIPTION
**Pull request by fjtrujy**
## Description

This PR fixes issues with `plists/inline-swift4.stencil` or `plists/inline-swift5.stencil` templates to generate the content of a list that contains arrays.

So far the content of the plist is:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>Names</key>
	<array>
		<string>Jhon</string>
		<string>Peter</string>
		<string>Nick</string>
	</array>
	<key>Surnames</key>
	<array>
		<string>Smith</string>
		<string>Jhonson</string>
		<string>Williams</string>
	</array>
</dict>
</plist>
```

And visually it looks:

Screenshot 2019-10-30 at 12 53 11

Then in the `.swiftgen.yml` I just have somethings as:

```yml
.........
plist:
    inputs: Precompile/OriginPlists/
    outputs:
        templateName: inline-swift4
        output: Generated/Plist.swift
.....
```

When I execute it, it generates wrong data:
```swift
internal static let names: [String] = [Jhon, Peter, Nick]
internal static let surnames: [String] = [Smith, Jhonson, Williams]
```

If you see it misses the `quotes` in every single element in the array, and in the plist, they are declared as `String`

Digging into the template I found where the issue "could be", I have created a solution, I tried locally and is working fine, but I don't know if this is breaking something else. Besides I see that the same logic is applied in several templates, so not sure if in these templates is broken as well.

The solution that I applied in the `plists/inline-swift4.stencil` was:

```diff
@@ -55,9 +55,9 @@ import Foundation
     Date(timeIntervalSinceReferenceDate: {{ value.timeIntervalSinceReferenceDate }})
   {% elif metadata.type == "Optional" %}
     nil
-  {% elif metadata.type == "Array" and metadata.element.items %}
-    [{% for itemMetadata in metadata.element.items %}
-      {% call valueBlock value[forloop.counter0] itemMetadata %}
+  {% elif metadata.type == "Array" and value %}
+    [{% for currentValue in value %}
+      {% call valueBlock currentValue metadata.element %}
       {% if not forloop.last %}, {% endif %}
     {% endfor %}]
   {% elif metadata.type == "Dictionary" %}
```

The new output is:

```swift
internal static let names: [String] = ["Jhon", "Peter", "Nick"]
internal static let surnames: [String] = ["Smith", "Jhonson", "Williams"]
```

🎉 🎉 🎉 🎉 🎉 🎉 🎉 

Here you have the reference to the specific github issue that I created time ago. https://github.com/SwiftGen/SwiftGen/issues/662

Thanks!

### CODE REVIEW

**Comment by SwiftGen-Eve** [comment]
Hey 👋 I'm Eve, the friendly bot watching over SwiftGen 🤖

Thanks a lot for your contribution!



---

Seems like everything is in order 👍 You did a good job here! 🤝

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by AliSoftware** [comment]
Thanks for the PR!

It seems you didn't update the tests accordingly (or didn't add a test to cover this case that we missed). Could you add one please?
Also don't forget to credit yourself in the CHANGELOG ;)

File: CHANGELOG.md [hunk 1]
```diff
@@ -54,6 +54,9 @@
 
 ### Bug Fixes
 
+* Plist: Fix issue with Array in the Inline templates.
```
Comment (SwiftGen-Eve) [inline CHANGELOG.md]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/SwiftGen/SwiftGen/blob/4481992932bf16d3ef02b2ada1135c1a4a5e7889/CHANGELOG.md#L57"></span>Line describing your entry should end with a period and 2 spaces.</td>
    </tr>
  </tbody>
</table>

**Comment by fjtrujy** [comment]
Hello,
@AliSoftware I will try to create a proper test to fix this issue.
Thanks!

**Comment by djbe** [comment]
- I actually started updating this PR, and noticed the same change could be applied to json/yaml inline templates.
- Then after generating code (and adding a test case), I also noticed that this fix actually breaks code generation for mixed arrays (`[Any]`).
- ~Furthermore, I tried testing the old template with the given test case, and it just works...~

~Added my findings to #662, will need more testing to see if we actually need a PR.~
Edit: I wasn't pressing the right test shortcut 🤦

**Comment by djbe** [comment]
Clarification for these latest changes:
- Old template assumed that there was always going to be `metadata.element.items[...]`, which isn't true for homogeneous arrays such as `[String]`
- The initial fix always passed along `metadata.element`, which isn't correct for mixed arrays, where we need the right metadata for each item.
- Correct solution is to try to use metadata for specific item, and fallback to general element metadata.

**Comment by fjtrujy** [comment]
Thanks @djbe for progressing with the work.
I was not able to finish it.
Thanks again

File: Tests/Fixtures/Resources/YAML/good/mapping.yaml [hunk 2]
```diff
@@ -4,3 +4,7 @@ foo:
   bar: banana
   baz: orange
 car: null
+names:
+  - John
+  - Peter
+  - Nick
```
Comment (AliSoftware) [inline Tests/Fixtures/Resources/YAML/good/mapping.yaml:10]: Not sure if this should be addressed by that PR or a separate one but I just realised that we might need to add test cases for strings containing (escaped) quotes like `- "Craig \"Hair Force One\" Federighi"` to see how our templates handle those… 

e.g. maybe we should at least use `##"raw string literals"##` in our stencil templates to limit the risk with those edge cases?

**Review by AliSoftware — COMMENTED** [review]
The fix for the particular problem looks good to me.

But I think we might have another issue (and likely not just with strings in arrays but with strings in general) when they contain escaped quotes…
(Feel free to decide if it's better to either fix this in this current PR or to split that in a separate one)

File: Tests/Fixtures/Resources/YAML/good/mapping.yaml (hunk 2 above)
Comment (djbe) [reply Tests/Fixtures/Resources/YAML/good/mapping.yaml:10]: Check #551, where we have issues with multiline strings. Whatever fix there should also escape quotes.

But yeah, that's for a separate PR.

Comment (AliSoftware) [reply Tests/Fixtures/Resources/YAML/good/mapping.yaml:10]: Ah cool, as long as that's tracked. Then let's merge this one and tackle the quote escapes and multilines issues separately via #551 👍

**Review by AliSoftware — APPROVED** [review]

**Pull request merged**